File: demos/pr-review-bot/agents/security-agent/agent.py
#!/usr/bin/env python3
"""
Security Agent for PR Review Bot

Analyzes code for security vulnerabilities using Gemini LLM.
Demonstrates a Python agent in the Parallax multi-language ecosystem.
"""

import asyncio
import json
import logging
import os
import re
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

# Add SDK to path for development
sdk_path = Path(__file__).parent.parent.parent.parent.parent / "packages" / "sdk-python" / "src"
sys.path.insert(0, str(sdk_path))

from parallax import ParallaxAgent, serve_agent

# Optional: Gemini integration
try:
    import google.generativeai as genai
    GEMINI_AVAILABLE = True
except ImportError:
    GEMINI_AVAILABLE = False
    print("Warning: google-generativeai not installed. Using fallback analysis.")

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


# Security patterns to check (fallback when no LLM)
SECURITY_PATTERNS = [
    {
        "pattern": r"eval\s*\(",
        "severity": "critical",
        "issue": "Use of eval() can lead to code injection",
        "suggestion": "Avoid eval() - use safer alternatives like JSON.parse() for data"
    },
    {
        "pattern": r"innerHTML\s*=",
        "severity": "high",
        "issue": "Direct innerHTML assignment can lead to XSS",
        "suggestion": "Use textContent or sanitize HTML before insertion"
    },
    {
        "pattern": r"SELECT.*FROM.*WHERE.*\+|SELECT.*FROM.*WHERE.*\$\{|SELECT.*FROM.*WHERE.*'.*'",
        "severity": "critical",
        "issue": "Potential SQL injection - query built with string concatenation",
        "suggestion": "Use parameterized queries or an ORM"
    },
    {
        "pattern": r"(password|api_key|secret|token)\s*=\s*['\"][^'\"]+['\"]",
        "severity": "critical",
        "issue": "Hardcoded credentials detected",
        "suggestion": "Use environment variables or a secrets manager"
    },
    {
        "pattern": r"fetch\s*\([^)]*\+|fetch\s*\([^)]*\$\{",
        "severity": "high",
        "issue": "URL constructed with user input - potential SSRF",
        "suggestion": "Validate and sanitize URLs before fetching"
    },
    {
        "pattern": r"document\.write\s*\(",
        "severity": "high",
        "issue": "document.write can be exploited for XSS",
        "suggestion": "Use DOM manipulation methods instead"
    },
    {
        "pattern": r"exec\s*\(|execSync\s*\(",
        "severity": "critical",
        "issue": "Command execution can lead to shell injection",
        "suggestion": "Avoid exec - use specific APIs for the task"
    }
]


class SecurityAgent(ParallaxAgent):
    """Agent that analyzes code for security vulnerabilities."""

    def __init__(self):
        super().__init__(
            agent_id="security-agent",
            name="Security Analyzer",
            capabilities=["security", "code-analysis", "vulnerability-detection"],
            metadata={
                "expertise": 0.9,
                "language": "python",
                "description": "Analyzes code for security vulnerabilities"
            }
        )

        # Initialize Gemini if available
        self.gemini_model = None
        if GEMINI_AVAILABLE:
            api_key = os.getenv("GEMINI_API_KEY")
            if api_key:
                genai.configure(api_key=api_key)
                self.gemini_model = genai.GenerativeModel("gemini-2.0-flash")
                logger.info("Gemini LLM initialized for security analysis")
            else:
                logger.warning("GEMINI_API_KEY not set - using pattern-based analysis")

    async def analyze(self, task: str, data: Optional[Any] = None) -> Tuple[Any, float]:
        """
        Analyze code for security vulnerabilities.

        Args:
            task: Description of the analysis task
            data: Dict containing 'code' to analyze

        Returns:
            Tuple of (result_dict, confidence)
        """
        code = ""
        if isinstance(data, dict):
            code = data.get("code", "")
        elif isinstance(data, str):
            code = data

        if not code:
            return ({
                "findings": [],
                "summary": "No code provided for analysis",
                "severity": "none",
                "reasoning": "Empty input"
            }, 0.5)

        # Try LLM analysis first
        if self.gemini_model:
            try:
                result, confidence = await self._analyze_with_llm(code)
                return (result, confidence)
            except Exception as e:
                logger.warning(f"LLM analysis failed, falling back to patterns: {e}")

        # Fallback to pattern-based analysis
        return self._analyze_with_patterns(code)

    async def _analyze_with_llm(self, code: str) -> Tuple[Dict, float]:
        """Analyze code using Gemini LLM."""
        prompt = f"""Analyze the following code for security vulnerabilities.

For each issue found, provide:
- severity: "critical", "high", "medium", or "low"
- issue: Brief description of the vulnerability
- line_hint: Approximate location in the code
- suggestion: How to fix it

Also assess your confidence in the analysis (0.0 to 1.0).

Respond in JSON format:
{{
  "findings": [
    {{"severity": "...", "issue": "...", "line_hint": "...", "suggestion": "..."}}
  ],
  "summary": "Overall security assessment",
  "overall_severity": "critical|high|medium|low|none",
  "confidence": 0.85,
  "reasoning": "Why you are this confident"
}}

Code to analyze:
```
{code}
```"""

        response = await asyncio.to_thread(
            self.gemini_model.generate_content,
            prompt
        )

        # Parse response
        text = response.text

        # Extract JSON from response (handle markdown code blocks)
        json_match = re.search(r'```(?:json)?\s*([\s\S]*?)\s*```', text)
        if json_match:
            text = json_match.group(1)

        try:
            result = json.loads(text)
            confidence = result.get("confidence", 0.75)

            # Validate confidence is in range
            confidence = max(0.0, min(1.0, float(confidence)))

            return (result, confidence)
        except json.JSONDecodeError:
            # If JSON parsing fails, create structured response from text
            return ({
                "findings": [],
                "summary": text[:500],
                "overall_severity": "unknown",
                "reasoning": "Could not parse structured response"
            }, 0.5)

    def _analyze_with_patterns(self, code: str) -> Tuple[Dict, float]:
        """Analyze code using regex patterns (fallback)."""
        findings = []

        for pattern_def in SECURITY_PATTERNS:
            matches = re.finditer(pattern_def["pattern"], code, re.IGNORECASE)
            for match in matches:
                # Find line number
                line_num = code[:match.start()].count('\n') + 1

                findings.append({
                    "severity": pattern_def["severity"],
                    "issue": pattern_def["issue"],
                    "line_hint": f"Line {line_num}",
                    "suggestion": pattern_def["suggestion"],
                    "matched_text": match.group(0)[:50]  # First 50 chars of match
                })

        # Determine overall severity
        severities = [f["severity"] for f in findings]
        if "critical" in severities:
            overall = "critical"
        elif "high" in severities:
            overall = "high"
        elif "medium" in severities:
            overall = "medium"
        elif "low" in severities:
            overall = "low"
        else:
            overall = "none"

        # Calculate confidence based on findings
        # Pattern matching is deterministic but may miss things
        base_confidence = 0.7
        if len(findings) > 0:
            confidence = min(0.85, base_confidence + len(findings) * 0.03)
        else:
            # No findings could mean clean code OR missed vulnerabilities
            confidence = 0.6

        summary = f"Found {len(findings)} potential security issue(s)" if findings else "No obvious security issues detected"

        return ({
            "findings": findings,
            "summary": summary,
            "overall_severity": overall,
            "reasoning": "Pattern-based analysis (LLM not available)",
            "analysis_method": "pattern_matching"
        }, confidence)


async def main():
    """Start the Security Agent."""
    agent = SecurityAgent()

    # Handle shutdown gracefully
    loop = asyncio.get_event_loop()

    def handle_shutdown():
        logger.info("Shutting down Security Agent...")
        asyncio.create_task(agent.shutdown())

    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, handle_shutdown)

    # Start serving
    port = int(os.getenv("AGENT_PORT", "50100"))
    await agent.serve(port)

    logger.info(f"Security Agent running on port {port}")
    logger.info("Press Ctrl+C to stop")

    # Wait for termination
    await agent.wait_for_termination()


if __name__ == "__main__":
    import signal
    asyncio.run(main())

File: demos/pr-review-bot/agents/style-agent/agent.ts
/**
 * Style/Quality Agent for PR Review Bot
 *
 * Analyzes code for style issues, complexity, and quality using Gemini LLM.
 * Demonstrates a TypeScript agent in the Parallax multi-language ecosystem.
 */

import { ParallaxAgent, serveAgent } from '@parallax/sdk-typescript';
import { GoogleGenerativeAI } from '@google/generative-ai';

// Style patterns to check (fallback when no LLM)
const STYLE_PATTERNS = [
  {
    pattern: /if\s*\([^)]+\)\s*\{\s*if\s*\([^)]+\)\s*\{\s*if/g,
    severity: 'high',
    issue: 'Deeply nested conditionals (3+ levels)',
    suggestion: 'Extract conditions into separate functions or use early returns'
  },
  {
    pattern: /function\s+\w+\s*\([^)]{100,}\)/g,
    severity: 'medium',
    issue: 'Function has too many parameters',
    suggestion: 'Consider using an options object or breaking into smaller functions'
  },
  {
    pattern: /\bany\b/g,
    severity: 'medium',
    issue: 'Use of "any" type reduces type safety',
    suggestion: 'Define proper types or use "unknown" with type guards'
  },
  {
    pattern: /console\.(log|warn|error)\s*\(/g,
    severity: 'low',
    issue: 'Console statements should be removed in production',
    suggestion: 'Use a proper logging library or remove before merging'
  },
  {
    pattern: /TODO|FIXME|HACK|XXX/gi,
    severity: 'low',
    issue: 'Unresolved TODO/FIXME comment',
    suggestion: 'Address the TODO or create a tracking issue'
  },
  {
    pattern: /^\s{40,}/gm,
    severity: 'medium',
    issue: 'Excessive indentation suggests complex nesting',
    suggestion: 'Refactor to reduce nesting depth'
  },
  {
    pattern: /\bvar\b/g,
    severity: 'medium',
    issue: 'Use of "var" instead of "let" or "const"',
    suggestion: 'Prefer "const" for immutable values, "let" for mutable'
  }
];

class StyleAgent extends ParallaxAgent {
  private gemini: GoogleGenerativeAI | null = null;
  private model: any = null;

  constructor() {
    super(
      'style-agent',
      'Code Style Analyzer',
      ['style', 'code-quality', 'complexity-analysis'],
      {
        expertise: 0.85,
        language: 'typescript',
        description: 'Analyzes code for style, quality, and complexity issues'
      }
    );

    // Initialize Gemini if API key is available
    const apiKey = process.env.GEMINI_API_KEY;
    if (apiKey) {
      this.gemini = new GoogleGenerativeAI(apiKey);
      this.model = this.gemini.getGenerativeModel({ model: 'gemini-2.0-flash' });
      console.log('Gemini LLM initialized for style analysis');
    } else {
      console.warn('GEMINI_API_KEY not set - using pattern-based analysis');
    }
  }

  async analyze(_task: string, data?: any): Promise<{
    value: any;
    confidence: number;
    reasoning?: string;
  }> {
    const code = typeof data === 'string' ? data : data?.code || '';

    if (!code) {
      return {
        value: {
          findings: [],
          summary: 'No code provided for analysis',
          quality_score: 0,
        },
        confidence: 0.5,
        reasoning: 'Empty input'
      };
    }

    // Try LLM analysis first
    if (this.model) {
      try {
        return await this.analyzeWithLLM(code);
      } catch (error) {
        console.warn('LLM analysis failed, falling back to patterns:', error);
      }
    }

    // Fallback to pattern-based analysis
    return this.analyzeWithPatterns(code);
  }

  private async analyzeWithLLM(code: string): Promise<{
    value: any;
    confidence: number;
    reasoning?: string;
  }> {
    const prompt = `Analyze the following code for style and quality issues.

For each issue found, provide:
- severity: "high", "medium", or "low"
- issue: Brief description of the problem
- line_hint: Approximate location in the code
- suggestion: How to improve it

Also provide:
- quality_score: 0-100 rating of overall code quality
- confidence: Your confidence in this analysis (0.0 to 1.0)

Focus on:
- Code complexity and nesting
- Naming conventions
- Function/method length
- Type safety
- Code organization
- Best practices

Respond in JSON format:
{
  "findings": [
    {"severity": "...", "issue": "...", "line_hint": "...", "suggestion": "..."}
  ],
  "summary": "Overall style assessment",
  "quality_score": 75,
  "confidence": 0.85,
  "reasoning": "Why you are this confident"
}

Code to analyze:
\`\`\`
${code}
\`\`\``;

    const result = await this.model.generateContent(prompt);
    const text = result.response.text();

    // Extract JSON from response
    const jsonMatch = text.match(/```(?:json)?\s*([\s\S]*?)\s*```/);
    const jsonText = jsonMatch ? jsonMatch[1] : text;

    try {
      const parsed = JSON.parse(jsonText);
      const confidence = Math.max(0, Math.min(1, parsed.confidence || 0.75));

      return {
        value: parsed,
        confidence,
        reasoning: parsed.reasoning || 'LLM analysis completed'
      };
    } catch {
      return {
        value: {
          findings: [],
          summary: text.substring(0, 500),
          quality_score: 50,
        },
        confidence: 0.5,
        reasoning: 'Could not parse structured response'
      };
    }
  }

  private analyzeWithPatterns(code: string): {
    value: any;
    confidence: number;
    reasoning?: string;
  } {
    const findings: Array<{
      severity: string;
      issue: string;
      line_hint: string;
      suggestion: string;
    }> = [];

    for (const pattern of STYLE_PATTERNS) {
      const matches = code.matchAll(pattern.pattern);
      for (const match of matches) {
        const lineNum = code.substring(0, match.index || 0).split('\n').length;
        findings.push({
          severity: pattern.severity,
          issue: pattern.issue,
          line_hint: `Line ${lineNum}`,
          suggestion: pattern.suggestion
        });
      }
    }

    // Calculate quality score
    const deductions = findings.reduce((sum, f) => {
      if (f.severity === 'high') return sum + 15;
      if (f.severity === 'medium') return sum + 8;
      return sum + 3;
    }, 0);

    const qualityScore = Math.max(0, 100 - deductions);

    // Confidence based on analysis coverage
    const baseConfidence = 0.65;
    const confidence = findings.length > 0
      ? Math.min(0.8, baseConfidence + findings.length * 0.02)
      : 0.55; // Lower confidence when no issues found (might have missed some)

    return {
      value: {
        findings,
        summary: findings.length > 0
          ? `Found ${findings.length} style issue(s)`
          : 'No obvious style issues detected',
        quality_score: qualityScore,
        analysis_method: 'pattern_matching'
      },
      confidence,
      reasoning: 'Pattern-based analysis (LLM not available)'
    };
  }
}

async function main() {
  const agent = new StyleAgent();
  const port = parseInt(process.env.AGENT_PORT || '50101', 10);

  await serveAgent(agent, port);

  console.log(`Style Agent running on port ${port}`);
  console.log('Press Ctrl+C to stop');
}

main().catch(console.error);
